File: LeetCode/1356/cpp/lc1356.cpp
#include <algorithm>
#include <cassert>
#include <vector>

class Solution {
public:
  std::vector<int> sortByBits(std::vector<int>& arr) {
    std::sort(arr.begin(), arr.end(), [](const int& lhs, const int& rhs){
              auto bl = __builtin_popcount(lhs);
              auto br = __builtin_popcount(rhs);
              return (bl == br) ? lhs < rhs : bl < br;
    });
    return arr;
  }
};

int main() {
  Solution s;

  std::vector<int> example1 = {0, 1, 2, 3, 4, 5, 6, 7, 8};
  std::vector<int> expected_ans1 = {0, 1, 2, 4, 8, 3, 5, 6, 7};
  assert(s.sortByBits(example1) == expected_ans1);

  std::vector<int> example2 = {1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1};
  std::vector<int> expected_ans2 = {1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024};
  assert(s.sortByBits(example2) == expected_ans2);
}

File: LeetCode/1356/golang/lc1356.go
package lc1356

import (
	"math/bits"
	"slices"
)

// Runtime complexity: O(sort)
// - (this is due to OnesCount64() being a constant time function)
// Auxiliary space: O(sort)
// Subjective level: easy
// Solved on: 2026-02-25
func sortByBits(arr []int) []int {
	slices.SortFunc(arr, func(lhs int, rhs int) int {
		bcl := bits.OnesCount64(uint64(lhs))
		bcr := bits.OnesCount64(uint64(rhs))
		if bcl == bcr {
			return lhs - rhs
		}
		return bcl - bcr
	})
	return arr
}
